logicalplan/filter.go

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"fmt"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+type FilteredSelector struct {
```

Comment:
So this seems to change PromQL spec and adds a new `filter()` function to it. I'm wondering how safe this is for example `filter` is implemented upstream, in which case it may do something completely different and we end up with diverging specs.

Reply:
That's true, we are adding a new AST node type, similar to how StepInvariant was added upstream: https://github.com/prometheus/prometheus/blob/96d5a32659f0e3928c10a771e50123fead9828bd/promql/parser/ast.go#L178-L183

However, this is not changing the actual spec of PromQL. It only changes the parsed AST. In the best case scenario we would have our own structs copied from the AST and work with them. I thought that might be an overkill for now so I decided to start with something simpler. 

If something like this was added to upsteram PromQL, I would expect some tests to fail somewhere :) I am not sure if there is a better way to prevent such incompatibilities, and I don't also see how exactly they would manifests.

Reply:
Ack! I see! This makes sense, thanks!

Reply:
Yes, we have to extend AST if we want to reuse it and avoid transforming types too many times.